File: apple/packages/EnteComponents/Sources/EnteComponents/Surfaces.swift
import SwiftUI

public enum BannerKind {
    case failure
    case information
    case success
    case warning
    case neutral
}

public struct FilterChip: View {
    @Environment(\.entePalette) private var palette
    private let label: String
    private let systemImage: String?
    private let selected: Bool
    private let enabled: Bool
    private let action: () -> Void

    public init(
        _ label: String,
        systemImage: String? = nil,
        selected: Bool,
        enabled: Bool = true,
        action: @escaping () -> Void
    ) {
        self.label = label
        self.systemImage = systemImage
        self.selected = selected
        self.enabled = enabled
        self.action = action
    }

    public var body: some View {
        Button(action: action) {
            HStack(spacing: EnteSpacing.sm) {
                if let systemImage {
                    Image(systemName: systemImage)
                        .frame(width: 16, height: 16)
                }
                Text(label)
                if selected {
                    Image(systemName: "xmark")
                        .font(.system(size: 14, weight: .medium))
                        .frame(width: 14, height: 14)
                }
            }
                .font(EnteTypography.mini)
                .lineLimit(1)
                .padding(.leading, systemImage == nil ? 18 : EnteSpacing.md)
                .padding(.trailing, selected ? EnteSpacing.md : (systemImage == nil ? 18 : EnteSpacing.lg))
                .padding(.vertical, EnteSpacing.md)
                .frame(minHeight: 40)
        }
        .buttonStyle(.plain)
        .foregroundStyle(foreground)
        .background(background)
        .clipShape(Capsule())
        .disabled(!enabled)
        .accessibilityAddTraits(selected ? .isSelected : [])
    }

    private var background: Color {
        selected ? (palette.isDark ? Color(hex: 0xF4F4F4) : Color(hex: 0x161616)) : palette.surface
    }

    private var foreground: Color {
        selected ? (palette.isDark ? .black : .white) : (enabled ? palette.mutedText : palette.disabledText)
    }
}

public struct Banner: View {
    @Environment(\.entePalette) private var palette
    private let title: String
    private let subtitle: String?
    private let kind: BannerKind
    private let systemImage: String?
    private let action: (() -> Void)?

    public init(
        _ title: String,
        subtitle: String? = nil,
        kind: BannerKind = .neutral,
        systemImage: String? = nil,
        action: (() -> Void)? = nil
    ) {
        self.title = title
        self.subtitle = subtitle
        self.kind = kind
        self.systemImage = systemImage
        self.action = action
    }

    public var body: some View {
        Group {
            if let action {
                Button(action: action) { content }
                    .buttonStyle(.plain)
            } else {
                content
            }
        }
        .background(palette.surface)
        .clipShape(RoundedRectangle(cornerRadius: EnteRadius.button, style: .continuous))
    }

    private var content: some View {
        HStack(spacing: EnteSpacing.lg) {
            if let systemImage {
                Image(systemName: systemImage)
                    .frame(width: 24, height: 24)
                    .foregroundStyle(accent)
            }
            VStack(alignment: .leading, spacing: EnteSpacing.xs) {
                Text(title)
                    .font(EnteTypography.bodyBold)
                    .foregroundStyle(kind == .neutral ? palette.text : accent)
                    .lineLimit(subtitle == nil ? 2 : 1)
                if let subtitle {
                    Text(subtitle)
                        .font(EnteTypography.mini)
                        .foregroundStyle(palette.mutedText)
                        .lineLimit(2)
                }
            }
            Spacer(minLength: 0)
        }
        .padding(.horizontal, EnteSpacing.lg)
        .padding(.vertical, EnteSpacing.md)
        .frame(maxWidth: .infinity, minHeight: 66, alignment: .leading)
    }

    private var accent: Color {
        switch kind {
        case .failure: palette.danger
        case .information: palette.information
        case .success, .neutral: palette.primaryDark
        case .warning: palette.caution
        }
    }
}

public struct Tag: View {
    @Environment(\.entePalette) private var palette
    private let label: String
    private let selected: Bool
    private let enabled: Bool
    private let action: () -> Void

    public init(
        _ label: String,
        selected: Bool,
        enabled: Bool = true,
        action: @escaping () -> Void
    ) {
        self.label = label
        self.selected = selected
        self.enabled = enabled
        self.action = action
    }

    public var body: some View {
        Button(action: action) {
            Text(label)
                .font(EnteTypography.body)
                .lineLimit(1)
                .padding(.horizontal, EnteSpacing.xl)
                .padding(.vertical, EnteSpacing.md)
                .frame(minHeight: 44)
        }
        .buttonStyle(.plain)
        .foregroundStyle(selected ? .white : (enabled ? palette.mutedText : palette.disabledText))
        .background(selected ? palette.primary : palette.surface)
        .clipShape(RoundedRectangle(cornerRadius: EnteRadius.large, style: .continuous))
        .disabled(!enabled)
        .accessibilityAddTraits(selected ? .isSelected : [])
    }
}

public struct MenuRow: View {
    @Environment(\.entePalette) private var palette
    private let title: String
    private let subtitle: String?
    private let systemImage: String?
    private let selected: Bool
    private let enabled: Bool
    private let action: () -> Void

    public init(
        _ title: String,
        subtitle: String? = nil,
        systemImage: String? = nil,
        selected: Bool = false,
        enabled: Bool = true,
        action: @escaping () -> Void
    ) {
        self.title = title
        self.subtitle = subtitle
        self.systemImage = systemImage
        self.selected = selected
        self.enabled = enabled
        self.action = action
    }

    public var body: some View {
        Button(action: action) {
            HStack(spacing: EnteSpacing.md) {
                if let systemImage {
                    Image(systemName: systemImage)
                        .frame(width: 36, height: 36)
                        .foregroundStyle(palette.mutedText)
                }
                VStack(alignment: .leading, spacing: EnteSpacing.xs) {
                    Text(title)
                        .font(EnteTypography.body)
                        .foregroundStyle(enabled ? palette.text : palette.disabledText)
                        .lineLimit(subtitle == nil ? 2 : 1)
                    if let subtitle {
                        Text(subtitle)
                            .font(EnteTypography.mini)
                            .foregroundStyle(palette.mutedText)
                            .lineLimit(2)
                    }
                }
                Spacer(minLength: 0)
            }
            .padding(.leading, systemImage == nil ? EnteSpacing.lg : EnteSpacing.md)
            .padding(.trailing, EnteSpacing.md)
            .padding(.vertical, 9)
            .frame(minHeight: 58)
        }
        .buttonStyle(.plain)
        .background(palette.surface)
        .overlay {
            RoundedRectangle(cornerRadius: EnteRadius.button, style: .continuous)
                .stroke(selected ? palette.border : .clear, lineWidth: 1)
        }
        .clipShape(RoundedRectangle(cornerRadius: EnteRadius.button, style: .continuous))
        .disabled(!enabled)
    }
}

public struct SettingsRow: View {
    @Environment(\.entePalette) private var palette
    private let title: String
    private let subtitle: String?
    private let systemImage: String?
    private let destructive: Bool
    private let action: () -> Void

    public init(
        _ title: String,
        subtitle: String? = nil,
        systemImage: String? = nil,
        destructive: Bool = false,
        action: @escaping () -> Void
    ) {
        self.title = title
        self.subtitle = subtitle
        self.systemImage = systemImage
        self.destructive = destructive
        self.action = action
    }

    public var body: some View {
        Button(action: action) {
            HStack(spacing: EnteSpacing.lg) {
                if let systemImage {
                    Image(systemName: systemImage)
                        .frame(width: 24, height: 24)
                        .foregroundStyle(destructive ? palette.danger : palette.mutedText)
                }
                VStack(alignment: .leading, spacing: 0) {
                    Text(title)
                        .font(EnteTypography.body)
                        .foregroundStyle(destructive ? palette.danger : palette.text)
                        .lineLimit(2)
                    if let subtitle {
                        Text(subtitle)
                            .font(EnteTypography.mini)
                            .foregroundStyle(palette.mutedText)
                            .lineLimit(1)
                    }
                }
                Spacer(minLength: 0)
                Image(systemName: "chevron.right")
                    .font(.system(size: 14, weight: .medium))
                    .foregroundStyle(palette.mutedText)
            }
            .padding(.vertical, EnteSpacing.md)
        }
        .buttonStyle(.plain)
    }
}

public struct MenuGroup<Content: View>: View {
    @Environment(\.entePalette) private var palette
    private let content: Content

    public init(@ViewBuilder content: () -> Content) {
        self.content = content()
    }

    public var body: some View {
        content
            .background {
                RoundedRectangle(cornerRadius: EnteRadius.button, style: .continuous)
                    .fill(palette.surface)
            }
            .clipShape(RoundedRectangle(cornerRadius: EnteRadius.button, style: .continuous))
    }
}

public struct EnteDivider: View {
    @Environment(\.entePalette) private var palette

    public init() {}

    public var body: some View {
        Divider().overlay(palette.faintBorder)
    }
}

public struct Avatar: View {
    @Environment(\.entePalette) private var palette
    private let name: String
    private let identity: String
    private let size: AvatarSize

    public init(_ name: String, identity: String? = nil, size: AvatarSize = .regular) {
        self.name = name
        self.identity = identity ?? name
        self.size = size
    }

    public var body: some View {
        Text(initials)
            .font(size.font)
            .foregroundStyle(Color.white)
            .frame(width: size.dimension, height: size.dimension)
            .background(colour)
            .clipShape(Circle())
            .overlay {
                Circle().stroke(palette.background, lineWidth: size.border)
            }
    }

    private var initials: String {
        let words = name.split(whereSeparator: { $0.isWhitespace })
        guard let first = words.first?.first else { return "?" }
        return String(first).uppercased() + (words.count > 1 ? String(words.last!.first!).uppercased() : "")
    }

    private var colour: Color {
        [palette.caution, palette.primary, Color(hex: 0xF24822), Color(hex: 0xDF61BB), Color(hex: 0x9610D6), Color(hex: 0x1071FF), Color(hex: 0x00B8D4)][identityHash % 7]
    }

    private var identityHash: Int {
        Int(identity.trimmingCharacters(in: .whitespacesAndNewlines).lowercased().utf8.reduce(UInt32(0x811c9dc5)) { ($0 ^ UInt32($1)) &* 0x01000193 })
    }
}

public enum AvatarSize {
    case extraSmall
    case small
    case regular
    case medium
    case large
    case contact

    var dimension: CGFloat {
        switch self {
        case .extraSmall: 16
        case .small: 20
        case .regular: 24
        case .medium: 28
        case .large: 32
        case .contact: 56
        }
    }

    var font: Font {
        switch self {
        case .extraSmall: EnteTypography.avatarExtraSmall
        case .small: EnteTypography.avatarSmall
        case .regular, .medium, .large: EnteTypography.mini
        case .contact: EnteTypography.heading2
        }
    }

    var border: CGFloat {
        switch self {
        case .large, .contact: 2
        default: 1
        }
    }
}

File: android/packages/components/src/main/java/io/ente/components/Surfaces.kt
package io.ente.components

import java.util.Locale
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.selection.toggleable
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.luminance
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp

enum class BannerKind { Failure, Information, Success, Warning, Neutral }

@Composable
fun FilterChip(
    label: String,
    selected: Boolean,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
    leading: @Composable (() -> Unit)? = null,
    trailing: @Composable (() -> Unit)? = null,
) {
    val palette = LocalEntePalette.current
    val background = if (selected) {
        if (palette.background.luminance() < 0.5f) Color(0xFFF4F4F4) else Color(0xFF161616)
    } else {
        palette.surface
    }
    val foreground = when {
        selected -> if (palette.background.luminance() < 0.5f) Color.Black else Color.White
        enabled -> palette.mutedText
        else -> palette.disabledText
    }
    val hasAutoTrailing = selected && trailing == null
    val startPadding = when {
        leading != null && trailing == null -> EnteSpacing.md
        trailing != null -> EnteSpacing.lg
        else -> 18.dp
    }
    val endPadding = when {
        leading != null && trailing == null -> if (selected) EnteSpacing.md else EnteSpacing.lg
        trailing != null || selected -> EnteSpacing.md
        else -> 18.dp
    }

    Surface(
        color = background,
        contentColor = foreground,
        shape = RoundedCornerShape(24.dp),
        modifier = modifier
            .heightIn(min = 40.dp)
            .toggleable(value = selected, enabled = enabled, role = Role.Checkbox) { onClick() },
    ) {
        Row(
            modifier = Modifier.padding(start = startPadding, end = endPadding, top = EnteSpacing.md, bottom = EnteSpacing.md),
            horizontalArrangement = Arrangement.Center,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            leading?.let {
                Box(Modifier.size(16.dp), contentAlignment = Alignment.Center) { it() }
                Spacer(Modifier.width(EnteSpacing.sm))
            }
            Text(label, maxLines = 1, overflow = TextOverflow.Ellipsis, style = EnteTypography.mini)
            if (trailing != null || hasAutoTrailing) {
                Spacer(Modifier.width(EnteSpacing.sm))
                if (trailing != null) {
                    Box(Modifier.size(16.dp), contentAlignment = Alignment.Center) { trailing() }
                } else {
                    RemoveIcon(foreground)
                }
            }
        }
    }
}

@Composable
fun Tag(
    label: String,
    selected: Boolean,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
    leading: @Composable (() -> Unit)? = null,
    trailing: @Composable (() -> Unit)? = null,
) {
    val palette = LocalEntePalette.current
    Surface(
        color = if (selected) palette.primary else palette.surface,
        contentColor = if (selected) Color.White else if (enabled) palette.mutedText else palette.disabledText,
        shape = RoundedCornerShape(EnteRadius.large),
        modifier = modifier
            .heightIn(min = 44.dp)
            .semantics { this.selected = selected }
            .clickable(enabled = enabled, role = Role.Button, onClick = onClick),
    ) {
        Row(
            modifier = Modifier.padding(
                start = if (leading != null) EnteSpacing.md else EnteSpacing.xl,
                end = if (trailing != null) EnteSpacing.md else EnteSpacing.xl,
                top = EnteSpacing.md,
                bottom = EnteSpacing.md,
            ),
            horizontalArrangement = Arrangement.Center,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            leading?.let {
                Box(Modifier.size(20.dp), contentAlignment = Alignment.Center) { it() }
                Spacer(Modifier.width(EnteSpacing.xs))
            }
            Text(label, maxLines = 1, overflow = TextOverflow.Ellipsis, style = EnteTypography.body)
            trailing?.let {
                Spacer(Modifier.width(EnteSpacing.xs))
                Box(Modifier.size(20.dp), contentAlignment = Alignment.Center) { it() }
            }
        }
    }
}

@Composable
fun MenuRow(
    title: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    subtitle: String? = null,
    selected: Boolean = false,
    enabled: Boolean = true,
    leading: @Composable (() -> Unit)? = null,
    trailing: @Composable (() -> Unit)? = null,
) {
    val palette = LocalEntePalette.current
    Surface(
        color = palette.surface,
        shape = RoundedCornerShape(EnteRadius.button),
        border = if (selected) androidx.compose.foundation.BorderStroke(1.dp, palette.border) else null,
        modifier = modifier
            .fillMaxWidth()
            .heightIn(min = 58.dp)
            .clickable(enabled = enabled, role = Role.Button, onClick = onClick),
    ) {
        Row(
            modifier = Modifier.padding(
                start = if (leading == null) EnteSpacing.lg else EnteSpacing.md,
                end = EnteSpacing.md,
                top = 9.dp,
                bottom = 9.dp,
            ),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            leading?.let {
                Box(Modifier.size(36.dp), contentAlignment = Alignment.Center) { it() }
                Spacer(Modifier.width(EnteSpacing.md))
            }
            Column(Modifier.weight(1f)) {
                Text(
                    title,
                    maxLines = if (subtitle == null) 2 else 1,
                    overflow = TextOverflow.Ellipsis,
                    style = EnteTypography.body,
                    color = if (enabled) palette.text else palette.disabledText,
                )
                subtitle?.let {
                    Spacer(Modifier.size(EnteSpacing.xs))
                    Text(it, maxLines = 2, overflow = TextOverflow.Ellipsis, style = EnteTypography.mini, color = palette.mutedText)
                }
            }
            trailing?.let {
                Spacer(Modifier.width(EnteSpacing.md))
                Box(Modifier.size(36.dp), contentAlignment = Alignment.Center) { it() }
            }
        }
    }
}

@Composable
fun MenuGroup(
    modifier: Modifier = Modifier,
    content: @Composable () -> Unit,
) {
    Surface(
        color = LocalEntePalette.current.surface,
        shape = RoundedCornerShape(EnteRadius.button),
        modifier = modifier.fillMaxWidth(),
    ) {
        Column(content = { content() })
    }
}

@Composable
fun Banner(
    title: String,
    onClick: (() -> Unit)? = null,
    modifier: Modifier = Modifier,
    kind: BannerKind = BannerKind.Neutral,
    subtitle: String? = null,
    leading: @Composable (() -> Unit)? = null,
    trailing: @Composable (() -> Unit)? = null,
) {
    val palette = LocalEntePalette.current
    val accent = when (kind) {
        BannerKind.Failure -> palette.danger
        BannerKind.Information -> palette.information
        BannerKind.Success, BannerKind.Neutral -> palette.primaryDark
        BannerKind.Warning -> palette.caution
    }
    val titleColor = if (kind == BannerKind.Neutral) palette.text else accent
    val interaction = if (onClick == null) Modifier else Modifier.clickable(role = Role.Button, onClick = onClick)

    Surface(
        color = palette.surface,
        shape = RoundedCornerShape(EnteRadius.button),
        modifier = modifier
            .fillMaxWidth()
            .defaultMinSize(minHeight = 66.dp)
            .then(interaction),
    ) {
        Row(
            modifier = Modifier.padding(horizontal = EnteSpacing.lg, vertical = EnteSpacing.md),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            leading?.let {
                Box(Modifier.size(24.dp), contentAlignment = Alignment.Center) { it() }
                Spacer(Modifier.width(EnteSpacing.lg))
            }
            Column(Modifier.weight(1f)) {
                Text(title, maxLines = if (subtitle == null) 2 else 1, overflow = TextOverflow.Ellipsis, style = EnteTypography.bodyBold, color = titleColor)
                subtitle?.let {
                    Spacer(Modifier.size(EnteSpacing.xs))
                    Text(it, maxLines = 2, overflow = TextOverflow.Ellipsis, style = EnteTypography.mini, color = palette.mutedText)
                }
            }
            trailing?.let {
                Spacer(Modifier.width(EnteSpacing.md))
                Box(Modifier.size(38.dp), contentAlignment = Alignment.Center) { it() }
            }
        }
    }
}

@Composable
fun SettingsRow(
    title: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    subtitle: String? = null,
    destructive: Boolean = false,
    leading: @Composable (() -> Unit)? = null,
    trailing: @Composable (() -> Unit)? = null,
) {
    val palette = LocalEntePalette.current
    val textColor = if (destructive) palette.danger else palette.text
    Row(
        modifier = modifier
            .fillMaxWidth()
            .clickable(role = Role.Button, onClick = onClick)
            .padding(vertical = EnteSpacing.md),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        leading?.let {
            Box(Modifier.size(24.dp), contentAlignment = Alignment.Center) { it() }
            Spacer(Modifier.width(EnteSpacing.lg))
        }
        Column(Modifier.weight(1f)) {
            Text(title, maxLines = 2, overflow = TextOverflow.Ellipsis, style = EnteTypography.body, color = textColor)
            subtitle?.let { Text(it, maxLines = 1, overflow = TextOverflow.Ellipsis, style = EnteTypography.mini, color = palette.mutedText) }
        }
        trailing?.let {
            Spacer(Modifier.width(EnteSpacing.md))
            Box(Modifier.size(24.dp), contentAlignment = Alignment.Center) { it() }
        }
    }
}

@Composable
fun EnteDivider(modifier: Modifier = Modifier) {
    HorizontalDivider(modifier, color = LocalEntePalette.current.faintBorder)
}

@Composable
fun Avatar(
    name: String,
    modifier: Modifier = Modifier,
    identity: String = name,
    size: AvatarSize = AvatarSize.Regular,
) {
    val palette = LocalEntePalette.current
    val color = avatarColors(palette)[avatarIndex(identity)]
    Box(
        modifier = modifier
            .size(size.dimension)
            .clip(CircleShape)
            .background(color)
            .border(size.border, palette.background, CircleShape),
        contentAlignment = Alignment.Center,
    ) {
        Text(avatarInitials(name), style = size.textStyle, color = Color.White)
    }
}

enum class AvatarSize(val dimension: androidx.compose.ui.unit.Dp, val border: androidx.compose.ui.unit.Dp, val textStyle: androidx.compose.ui.text.TextStyle) {
    ExtraSmall(16.dp, 1.dp, EnteTypography.avatarExtraSmall),
    Small(20.dp, 1.dp, EnteTypography.avatarSmall),
    Regular(24.dp, 1.dp, EnteTypography.mini),
    Medium(28.dp, 1.dp, EnteTypography.mini),
    Large(32.dp, 2.dp, EnteTypography.mini),
    Contact(56.dp, 2.dp, EnteTypography.heading2),
}

private fun avatarColors(palette: Palette) = listOf(
    palette.caution, palette.primary, Color(0xFFF24822), Color(0xFFDF61BB),
    Color(0xFF9610D6), Color(0xFF1071FF), Color(0xFF00B8D4),
)

@Composable
private fun RemoveIcon(color: Color) {
    Canvas(Modifier.size(14.dp)) {
        val inset = size.width * 0.24f
        drawLine(color, Offset(inset, inset), Offset(size.width - inset, size.height - inset), 1.5.dp.toPx(), StrokeCap.Round)
        drawLine(color, Offset(size.width - inset, inset), Offset(inset, size.height - inset), 1.5.dp.toPx(), StrokeCap.Round)
    }
}

private fun avatarIndex(identity: String): Int {
    var hash = 0x811c9dc5L
    identity.trim().lowercase(Locale.ROOT).toByteArray(Charsets.UTF_8).forEach { byte ->
        hash = (hash xor (byte.toLong() and 0xff)) * 0x01000193L and 0xffffffffL
    }
    return (hash % 7).toInt()
}

private fun avatarInitials(name: String): String {
    val words = name.trim().split(Regex("\\s+")).filter(String::isNotEmpty)
    if (words.isEmpty()) return "?"
    return (words.first().first().uppercase() + words.drop(1).lastOrNull()?.first()?.uppercase().orEmpty())
}
